### PR TITLE
DOC: Blogpost_KD_Week (2 - 6)

### PR DESCRIPTION
**Pull request by deka27**
Summary:
This pull request introduces a blog post for week 2 - 6 of GSOC. It aims to provide insights and reflections for future work as well.

Key Changes:

Added a new blog post file in the posts/2024 directory.

### CODE REVIEW

**Review by robinroy03 — APPROVED** [review]
LGTM

**Comment by itellaetxe** [comment]
I cannot explicitly review, but LGTM

**Review by WassCodeur — APPROVED** [review]
LGTM, Great work

File: posts/2024/2024_06_12_kaustav_week_2.rst [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+     :category: gsoc
+
+Greetings, everyone!
+It's time for another update on my Google Summer of Code (GSoC) journey with DIPY. The second week of the Coding phase has been equally productive and exciting, with significant advancements in both tasks.
+
+Decorator Function Refinement
+----------------------------
+
+Building upon the initial implementation from the previous week, I have invested significant efforts in refining and correcting the decorator function.
+With the guidance of my mentor, Serge Koudoro, I have meticulously reviewed the code, identified areas for improvement, and made necessary changes to enhance its performance, readability, and maintainability.
```
Comment (skoudoro) [inline posts/2024/2024_06_12_kaustav_week_2.rst:16]: link to github for name

**Review by skoudoro — CHANGES_REQUESTED** [review]
Hi @deka27,

Thank you for your blogpost. For the next ones, It would be great if you could more technical and more personalized. Currently, I find the blogpost very generic. 


@WassCodeur @robinroy03 , I see LGTM but have you even try to build the blog post locally ? Do you know how to build the blog locally? Please this is a minimum. 

 I would like to see progress on reviewing code.  waiting for a second review and waiting for your feedback

File: posts/2024/2024_06_12_kaustav_week_2.rst (hunk 1 above)
Comment (deka27) [reply posts/2024/2024_06_12_kaustav_week_2.rst:16]: Okay, will do this.

Comment (WassCodeur) [reply posts/2024/2024_06_12_kaustav_week_2.rst:16]: Sure

File: posts/2024/2024_06_12_kaustav_week_2.rst [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+Documentation Updates
+---------------------
+
+To ensure the long-term maintainability and usability of the decorator, I have updated the necessary documentation with Numpy standard.
+This includes clear explanations of the decorator's functionality, guidelines for its usage, and examples to assist other contributors and users in leveraging its capabilities effectively.
+
+Lazy Loading Research
+---------------------
+
+In addition to decorators, I looked and tested some implementations of lazy loading in various libraries.
```
Comment (WassCodeur) [inline posts/2024/2024_06_12_kaustav_week_2.rst:33]: Could you perhaps specify in which library you tested this?

File: posts/2024/2024_06_12_kaustav_week_2.rst [hunk 3]
```diff
@@ -0,0 +1,42 @@
+My Journey Continues: Week 2 Progress with DIPY
```
Comment (robinroy03) [inline posts/2024/2024_06_12_kaustav_week_2.rst:1]: `Warning, treated as error:
/home/robin/Desktop/l/dipy.org/posts/2024/2024_06_12_kaustav_week_2.rst:13:Title underline too short.`

Can be fixed by updating line-13 with an extra `-`.

**Review by robinroy03 — CHANGES_REQUESTED** [review]
Hi @deka27, please see the below changes.

Also, the order specified for a PSF blog is as follows ([source](https://social.python-gsoc.org/@terri/112638715402005382))

1. What did you do this past week?
2. What is coming up next?
3. Did you get stuck anywhere?

**Comment by skoudoro** [comment]
Also, When you update your blog post and address the comments above, Can you rebase this PR? Thank you

File: posts/2024/2024_06_12_kaustav_week_2.rst [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+---------------------
+
+In addition to decorators, I looked and tested some implementations of lazy loading in various libraries.
+
+Final Thoughts
+--------------
+
+The second week of the Coding phase has been dedicated to refining and correcting the decorator function, laying a solid foundation for its seamless integration into DIPY's codebase. 
+
+Stay tuned for more updates as I continue to work on enhancing DIPY!
```
Comment (itellaetxe) [inline posts/2024/2024_06_12_kaustav_week_2.rst:40]: I think you need to also include what are the plans for next week, and where did you get stuck, complying with the [Python GSoC Blog requirements](https://python-gsoc.org/blogs.html).
The blog only covers the first point of "what did you do this week"

You don't necessarily need to dedicate specific sections for the info that is lacking, but it should show somewhere at least.

File: posts/2024/2024_06_19_kaustav_week_3.rst [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+Final Thoughts
+--------------
+
+The third week of the Coding phase has been a significant milestone, with the decorator implementation across multiple modules and the demonstration of lazy loading capabilities.
+I am grateful for the guidance and support provided by my mentor and the DIPY community, which have been instrumental in driving this project forward.
+
+Stay tuned for more updates as I continue to work on enhancing DIPY!
+
+Thank you for reading!
```
Comment (itellaetxe) [inline posts/2024/2024_06_19_kaustav_week_3.rst:41]: Same as in week 2, you need to include info about what is coming up next week, and where you got stuck.

**Review by itellaetxe — COMMENTED** [review]
Hi @deka27 sorry for the late review, I was unable to request my own review here. Now I can.

Overall, it is concise, but in my first "review comment" I did not realize about some lacking information. Could you include it, please? Just to comply with the Python GSoC blog requirements.

File: posts/2024/2024_06_12_kaustav_week_2.rst [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+.. post:: June 12 2024
+     :author: Kaustav
+     :tags: google
+     :category: gsoc
+
+Greetings, everyone!
+It's time for another update on my Google Summer of Code (GSoC) journey with DIPY. The second week of the Coding phase has been equally productive and exciting, with significant advancements in both tasks.
+
+Decorator Function Refinement
```
[the remark this answers is not in the document]
Comment (deka27) [reply posts/2024/2024_06_12_kaustav_week_2.rst:12]: Fixed.

File: posts/2024/2024_06_12_kaustav_week_2.rst (hunk 3 above)
Comment (deka27) [reply posts/2024/2024_06_12_kaustav_week_2.rst:1]: Fixed

File: posts/2024/2024_06_19_kaustav_week_3.rst [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+Parallel to the decorator implementation, I have also made progress on the lazy loading integration task.
+As a proof of concept, I have demonstrated a simple implementation of lazy loading in the align module of DIPY.
+This initial implementation showcases the potential benefits of lazy loading, such as improved memory efficiency and performance optimization.
+By delaying the initialization of objects until they are actually needed, lazy loading can significantly reduce the memory footprint of the application and enhance overall efficiency.
+
+Next Week
+---------
+
+The plan next week is to work on the Lazy Loading, implement in DIPY and
```
Comment (pjsjongsung) [inline posts/2024/2024_06_19_kaustav_week_3.rst:47]: The sentence here seems to be incomplete

File: posts/2024/2024_06_30_kaustav_week_4.rst [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+- Identifying key modules and functions that would benefit most from lazy loading.
+- Developing a strategy for implementing lazy loading without disrupting existing functionality.
+- Beginning the actual implementation process, by implementing for Sub-Modules of DIPY and sharing a build with my mentor for POC.
+- Preparing comprehensive tests to ensure the lazy loading implementation works as expected and doesn't introduce any bugs.
+
+Final Thoughts
+--------------
+
+While this week didn't see as much tangible progress as I would have liked, it has been an opportunity for reflection and preparation. I'm grateful for the support of my mentor and excited to return to full speed at the earliest.
```
Comment (robinroy03) [inline posts/2024/2024_06_30_kaustav_week_4.rst:38]: come back stronger :zap:

File: posts/2024/2024_06_12_kaustav_week_2.rst [hunk 9]
```diff
@@ -0,0 +1,56 @@
+My Journey Continues: Week 2 Progress with DIPY
+===============================================
+
+.. post:: June 12 2024
```
Comment (robinroy03) [inline posts/2024/2024_06_12_kaustav_week_2.rst:4]: The `post` date is the date when you wrote the blog. I believe you did both 2 and 3 on the same day.

**Review by robinroy03 — COMMENTED** [review]
Hi Kaustav, good work here!

The 3 blogs now answer the main [questions](https://social.python-gsoc.org/@terri/112638715402005382).
Check the below comment regarding `post` date.

File: posts/2024/2024_06_30_kaustav_week_4.rst [hunk 10]
```diff
@@ -0,0 +1,39 @@
+My Journey Continues: Week 4 Progress with DIPY
+===============================================
+
+.. post:: June 30 2024
```
Comment (skoudoro) [inline posts/2024/2024_06_30_kaustav_week_4.rst:4]: Week 4 should be June 23 and week 5 is June 30

**Review by skoudoro — CHANGES_REQUESTED** [review]
Hi @deka27,

Can provide week 5, address all comments and update week 4 which has the wrong date. Thank you

File: posts/2024/2024_07_07_kaustav_week_6.rst [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+     :tags: google
+     :category: gsoc
+
+Greetings, everyone!
+The sixth week of GSOC has been a hectic one. A ton of time in correcting errors and fixing PRs. 
+
+Decorator PR Fixes
+------------------
+
+With Decorator Implementation PRs in place, and the integration of my original PR for keyword (https://github.com/dipy/dipy/pull/3239), it was now time for the PRs which had decorator implementation for specific module to be integrated.
```
Comment (skoudoro) [inline posts/2024/2024_07_07_kaustav_week_6.rst:15]: why all this capital letter?

File: posts/2024/2024_07_07_kaustav_week_6.rst [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+Lazy Loading PR
+---------------
+
+I have submitted a PR for the Lazy Loading feature using the lazy_loader package. (https://github.com/dipy/dipy/pull/3288)
```
Comment (skoudoro) [inline posts/2024/2024_07_07_kaustav_week_6.rst:20]: link to `lazy_loader` package

File: posts/2024/2024_07_07_kaustav_week_6.rst [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+Difficulties & Challenges
+-------------------------
+
+I had and have a ton of Challenges ahead as I am having trouble with some PR, I am specifically having a challenge in completing:
```
Comment (skoudoro) [inline posts/2024/2024_07_07_kaustav_week_6.rst:25]: why capital to challenges?

File: posts/2024/2024_07_07_kaustav_week_6.rst [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+1. https://github.com/dipy/dipy/pull/3257
+- This PR, has issues related to `wls_fit_tensor` function, after the application of the decorator, I am receiving an error where it is telling me hat 3 positional arguments are allowed but 4 are being passed, but when I check the main function parameters, there are only 3 positional params, I am not able to figure why 1 extra arg is being passed.
```
Comment (skoudoro) [inline posts/2024/2024_07_07_kaustav_week_6.rst:27]: - too many comas
- `hat`? or `that` ?
- figure  ? or `figure out` 
- `arg` to stylized

File: posts/2024/2024_07_07_kaustav_week_6.rst [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+2. https://github.com/dipy/dipy/pull/3254
+- This PR, I have a question about the application of arguments separator, I am just awaiting response on applying the separator.
+3. https://github.com/dipy/dipy/pull/3288
+- This PR is my lazy loading PR, I am having trouble with the application of `type_stub` as the package is not able to recognise it and thus I am receiving error in importing the modules correctly. I was trying to make it as close as possible to scikit-image implementation, it hasn't happened yet. But I'm hopeful.
```
Comment (skoudoro) [inline posts/2024/2024_07_07_kaustav_week_6.rst:31]: recognize instead (american style)

**Review by skoudoro — COMMENTED** [review]
Hi @deka27, see below few comments.

it would be great if @WassCodeur, @itellaetxe  and @robinroy03 could have a look also.

Thanks

File: posts/2024/2024_07_07_kaustav_week_6.rst [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+------------------
+
+With Decorator Implementation PRs in place, and the integration of my original PR for keyword (https://github.com/dipy/dipy/pull/3239), it was now time for the PRs which had decorator implementation for specific module to be integrated.
+
+Lazy Loading PR
+---------------
+
+I have submitted a PR for the Lazy Loading feature using the lazy_loader package. (https://github.com/dipy/dipy/pull/3288)
+
+Difficulties & Challenges
```
Comment (robinroy03) [inline posts/2024/2024_07_07_kaustav_week_6.rst:22]: Add a `\n` between challenges 1, 2 & 3.

image

**Review by robinroy03 — COMMENTED** [review]
Good work Kaustav. I'm confident you'll fix the PR challenges you've got right now.

Please see the comment below for the week 6 blog review.

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 14 above)
Comment (itellaetxe) [inline posts/2024/2024_07_07_kaustav_week_6.rst:27]: ...has issues... --> ...**it** has issues...

...related to... --> ...related to **the**...

arg --> **argument**

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 15 above)
Comment (itellaetxe) [inline posts/2024/2024_07_07_kaustav_week_6.rst:31]: ...receiving an error in importing... --> ...receiving an error **when** importing...

...as possible to scikit-image implementation... --> ...as possible to **the** `scikit-image` implementation

File: posts/2024/2024_07_07_kaustav_week_6.rst [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+- This PR, has issues related to `wls_fit_tensor` function, after the application of the decorator, I am receiving an error where it is telling me hat 3 positional arguments are allowed but 4 are being passed, but when I check the main function parameters, there are only 3 positional params, I am not able to figure why 1 extra arg is being passed.
+2. https://github.com/dipy/dipy/pull/3254
+- This PR, I have a question about the application of arguments separator, I am just awaiting response on applying the separator.
+3. https://github.com/dipy/dipy/pull/3288
+- This PR is my lazy loading PR, I am having trouble with the application of `type_stub` as the package is not able to recognise it and thus I am receiving error in importing the modules correctly. I was trying to make it as close as possible to scikit-image implementation, it hasn't happened yet. But I'm hopeful.
+
+Next Week
+---------
+
+Next week I am planning to push through the remaining PRs of decorator implementation. Also for lazy_loader, I am planning to fix the error.
```
Comment (itellaetxe) [inline posts/2024/2024_07_07_kaustav_week_6.rst:36]: Style the word 'lazy_loader' --> `lady_loader`

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 12 above)
Comment (itellaetxe) [reply posts/2024/2024_07_07_kaustav_week_6.rst:20]: Style the word 'lazy_loader' --> `lazy_loader`

**Review by itellaetxe — COMMENTED** [review]
Good work Kaustav! Looking good. Just small grammar improvements/suggestions, see below:

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 15 above)
Comment (deka27) [reply posts/2024/2024_07_07_kaustav_week_6.rst:31]: Done. Thanks.

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 14 above)
Comment (deka27) [reply posts/2024/2024_07_07_kaustav_week_6.rst:27]: Done. Thanks.

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (deka27) [reply posts/2024/2024_07_07_kaustav_week_6.rst:15]: Done. Thanks.

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 16 above)
Comment (deka27) [reply posts/2024/2024_07_07_kaustav_week_6.rst:22]: Done. Thanks.

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 15 above)
Comment (deka27) [reply posts/2024/2024_07_07_kaustav_week_6.rst:31]: Done. Thanks.

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 14 above)
Comment (deka27) [reply posts/2024/2024_07_07_kaustav_week_6.rst:27]: Done. Thanks.
I changed arg to argument and highlighted it.

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 13 above)
Comment (deka27) [reply posts/2024/2024_07_07_kaustav_week_6.rst:25]: Fixed. Thanks.

File: posts/2024/2024_07_07_kaustav_week_6.rst (hunk 11 above)
Comment (deka27) [reply posts/2024/2024_07_07_kaustav_week_6.rst:15]: Fixed. Thanks.

**Review by skoudoro — APPROVED** [review]
Ok, thanks for the update, I will not wait @WassCodeur review and go ahead and merge.

**Pull request merged**